src/main/jbake/content/blog/2016/04/22/fully-non-blocking-cassandra-resultset-traversal.adoc

```diff
@@ -0,0 +1,318 @@
+= Fully non-blocking Cassandra ResultSet traversal with RxJava
+Thomas Segismont
+2016-04-22
+:icons: font
+:jbake-type: post
+:jbake-status: published
+:jbake-tags: blog, hawkular, rxjava, cassandra, asynchronous, reactive, backpressure
+
+In Hawkular Metrics, as we try to minimize resources needed to handle a massive number of clients,
+we avoid calling blocking APIs whenever possible.
+
+== Decorating Session methods with RxJava
+
+The DataStax Java driver offers methods to execute statements asynchronously. To make this work nicely with RxJava, we
+have a `Session` wrapper in our repository footnote:[https://git.io/vwGfw], exposing methods like the following:
+
+[source,java]
+.RxSession
+----
+/**
+ * Asynchronously execute a query and emit the corresponding {@link ResultSet} on the
+ * {@link Schedulers#computation()} scheduler.
+ *
+ * @param statement the CQL query to execute, of any {@link Statement} type
+ *
+ * @return an {@link Observable} emitting just one {@link ResultSet} item
+ */
+Observable<ResultSet> execute(Statement statement);
+----
+
+The implementation is quite straightforward, it simply converts the `ListenableFuture` returned by `Session#executeAsync`
+to an `Observable` with https://github.com/ReactiveX/RxJavaGuava[RxJavaGuava].
+
+[source,java]
+.RxSession implementation
+----
+@Override
+public Observable<ResultSet> execute(Statement statement) {
+    ResultSetFuture future = session.executeAsync(statement);
+    return ListenableFutureObservable.from(future, Schedulers.computation());
+}
+----
+
+IMPORTANT: Notice here that the `ResultSet` item will be emitted on the computation scheduler. This is
+what we want for data manipulation operations.
+
+From there we can use the usual methods of `Observable` to transform the `ResultSet`. For example, if we want to
+determine whether an insertion statement went well:
+
+[source,java]
+.Checking if insertion went well
+----
+Observable<Boolean> wasApplied = rxSession.execute(insertStatement).map(ResultSet::wasApplied);
+----
+
```

Comment:
2016-04-21 16:31 GMT+02:00 jsanda <notifications@github.com>:

> This is not the intended use of the wasApplied method. It is intended for
> conditional updates, i.e., light weight transactions. It always returns
> true for regular queries, read or write. In the case of a regular query,
> there is no need to call this method.
>


Thanks for pointing this out. I fixed the description.